include/math/utils/ndop.h

```diff
@@ -0,0 +1,221 @@
+        assert(index < get_num_planes());
+        return index >= get_num_axises() ? index - get_num_axises() : index + get_num_axises();
+    }
+
+    constexpr static math::float3 get_axis(size_t index) noexcept {
+        assert(index < axises.size());
+        return axises[index];
+    }
+
+    constexpr static math::float3 get_plane_normal(size_t index) noexcept {
+        assert(index < get_num_planes());
+        return index >= get_num_axises() ? -axises[index - get_num_axises()] : axises[index];
+     }
+
+private:
+    constexpr static auto sqrt3 = std::numbers::inv_sqrt3_v<float>;
+    constexpr static std::array<math::float3, 7> axises {
+        math::float3{1.0f, 0.0f, 0.0f},
+        math::float3{0.0f, 1.0f, 0.0f},
+        math::float3{0.0f, 0.0f, 1.0f},
+        math::float3{sqrt3, sqrt3, sqrt3},
+        math::float3{sqrt3, sqrt3, -sqrt3},
+        math::float3{sqrt3, -sqrt3, sqrt3},
+        math::float3{-sqrt3, sqrt3, sqrt3},
+    };
+
+    //static_assert(math::approx_equal(math::len_squared(axises[i]), 1.0f));
+};
+
+//template <auto Traits>
+class DOP : DOP14_Traits {
+   public:
+    constexpr explicit DOP() noexcept = default;
+
+    template <typename Iterator, typename UnaryFn>
+    constexpr static DOP compute(
+        Iterator vertices_begin, Iterator vertices_end, UnaryFn&& fn_get_position = std::identity{}) noexcept;
+
+    std::vector<math::float3> get_points() const noexcept;
+    constexpr std::optional<math::float3> get_center() const noexcept;
+    constexpr bool is_empty() const noexcept;
+    constexpr math::float3 get_size() const noexcept;
+
+    DOP transformed(const math::float4x4& transform) const noexcept;
+    constexpr DOP unite_with(const DOP& rhv) const noexcept;
+    constexpr DOP unite_with(const math::float3& point) const noexcept;
+
+    constexpr void expand (const math::float3& point) noexcept;
+    
+    constexpr bool contains(const math::float3& point) const noexcept;
+
+   private:
+    std::array<float, get_num_planes()> distances_ 
+    {
+        [](){
+            std::array<float, get_num_planes()> result;
+            std::fill(std::begin(result), std::end(result), std::numeric_limits<float>::lowest());
+            return result;
+        }()
+    };
+};
+
+constexpr std::optional<math::float3> DOP::get_center() const noexcept {
+    return is_empty() ? std::nullopt : std::optional{math::float3{std::midpoint(distances_[0], -distances_[get_inverse_plane_index(0)]),
+                                                      std::midpoint(distances_[1], -distances_[get_inverse_plane_index(1)]),
+                                                      std::midpoint(distances_[2], -distances_[get_inverse_plane_index(2)]),
+    }};
+}
+
+constexpr bool DOP::is_empty() const noexcept { 
+    return std::any_of(std::begin(distances_), std::end(distances_), [](const auto& d) { return d <= std::numeric_limits<float>::lowest(); });
```

Comment:
I think it could result in a false-positive if a DOP was created from points lying on a plane/line.

Reply:
Nah it should be fine